File: src/renderer/vulkan/Barrier.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "Barrier.h"

#include "Image.h"

#include <core/Assert.h>

namespace Reaper
{
VkImageMemoryBarrier2 get_vk_image_barrier(VkImage handle, const GPUTextureSubresource& subresource,
                                           GPUTextureAccess src, GPUTextureAccess dst, u32 src_queue_family_index,
                                           u32 dst_queue_family_index)
{
    const VkImageSubresourceRange view_range = GetVulkanImageSubresourceRange(subresource);

    return VkImageMemoryBarrier2{.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER_2,
                                 .pNext = nullptr,
                                 .srcStageMask = src.stage_mask,
                                 .srcAccessMask = src.access_mask,
                                 .dstStageMask = dst.stage_mask,
                                 .dstAccessMask = dst.access_mask,
                                 .oldLayout = src.image_layout,
                                 .newLayout = dst.image_layout,
                                 .srcQueueFamilyIndex = src_queue_family_index,
                                 .dstQueueFamilyIndex = dst_queue_family_index,
                                 .image = handle,
                                 .subresourceRange = view_range};
}

VkBufferMemoryBarrier2 get_vk_buffer_barrier(VkBuffer handle, const GPUBufferView& view, GPUBufferAccess src,
                                             GPUBufferAccess dst, u32 src_queue_family_index,
                                             u32 dst_queue_family_index)
{
    Assert(view.size_bytes > 0, "Invalid view size");

    return VkBufferMemoryBarrier2{.sType = VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER_2,
                                  .pNext = nullptr,
                                  .srcStageMask = src.stage_mask,
                                  .srcAccessMask = src.access_mask,
                                  .dstStageMask = dst.stage_mask,
                                  .dstAccessMask = dst.access_mask,
                                  .srcQueueFamilyIndex = src_queue_family_index,
                                  .dstQueueFamilyIndex = dst_queue_family_index,
                                  .buffer = handle,
                                  .offset = view.offset_bytes,
                                  .size = view.size_bytes};
}

VkMemoryBarrier2 get_vk_memory_barrier(GPUMemoryAccess src, GPUMemoryAccess dst)
{
    return VkMemoryBarrier2{
        .sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER_2,
        .pNext = nullptr,
        .srcStageMask = src.stage_mask,
        .srcAccessMask = src.access_mask,
        .dstStageMask = dst.stage_mask,
        .dstAccessMask = dst.access_mask,
    };
}

VkDependencyInfo get_vk_image_barrier_depency_info(std::span<const VkImageMemoryBarrier2> barriers)
{
    Assert(!barriers.empty());

    return VkDependencyInfo{.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO,
                            .pNext = nullptr,
                            .dependencyFlags = VK_FLAGS_NONE,
                            .memoryBarrierCount = 0,
                            .pMemoryBarriers = nullptr,
                            .bufferMemoryBarrierCount = 0,
                            .pBufferMemoryBarriers = nullptr,
                            .imageMemoryBarrierCount = static_cast<u32>(barriers.size()),
                            .pImageMemoryBarriers = barriers.data()};
}

VkDependencyInfo get_vk_buffer_barrier_depency_info(std::span<const VkBufferMemoryBarrier2> barriers)
{
    Assert(!barriers.empty());

    return VkDependencyInfo{.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO,
                            .pNext = nullptr,
                            .dependencyFlags = VK_FLAGS_NONE,
                            .memoryBarrierCount = 0,
                            .pMemoryBarriers = nullptr,
                            .bufferMemoryBarrierCount = static_cast<u32>(barriers.size()),
                            .pBufferMemoryBarriers = barriers.data(),
                            .imageMemoryBarrierCount = 0,
                            .pImageMemoryBarriers = nullptr};
}

VkDependencyInfo get_vk_memory_barrier_depency_info(std::span<const VkMemoryBarrier2> barriers)
{
    Assert(!barriers.empty());

    return VkDependencyInfo{.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO,
                            .pNext = nullptr,
                            .dependencyFlags = VK_FLAGS_NONE,
                            .memoryBarrierCount = static_cast<u32>(barriers.size()),
                            .pMemoryBarriers = barriers.data(),
                            .bufferMemoryBarrierCount = 0,
                            .pBufferMemoryBarriers = nullptr,
                            .imageMemoryBarrierCount = 0,
                            .pImageMemoryBarriers = nullptr};
}

} // namespace Reaper

File: src/renderer/vulkan/barrier.zig
// Port of src/renderer/vulkan/Barrier.h + Barrier.cpp
//
// Only the image-barrier half is needed so far; buffer and memory barriers
// arrive with the frame graph.

const vk = @import("vulkan");

pub const GPUTextureAccess = struct {
    stage_mask: vk.PipelineStageFlags2,
    access_mask: vk.AccessFlags2,
    image_layout: vk.ImageLayout,
};

pub const GPUBufferAccess = struct {
    stage_mask: vk.PipelineStageFlags2,
    access_mask: vk.AccessFlags2,
};

pub const GPUMemoryAccess = struct {
    stage_mask: vk.PipelineStageFlags2,
    access_mask: vk.AccessFlags2,
};

pub const GPUTextureSubresource = struct {
    aspect: vk.ImageAspectFlags,
    mip_offset: u32,
    mip_count: u32,
    layer_offset: u32,
    layer_count: u32,
};

/// Mirrors default_texture_subresource_one_color_mip().
pub fn defaultTextureSubresourceOneColorMip() GPUTextureSubresource {
    return .{
        .aspect = .{ .color_bit = true },
        .mip_offset = 0,
        .mip_count = 1,
        .layer_offset = 0,
        .layer_count = 1,
    };
}

pub fn getImageBarrier(
    handle: vk.Image,
    subresource: GPUTextureSubresource,
    src: GPUTextureAccess,
    dst: GPUTextureAccess,
    src_queue_family_index: u32,
    dst_queue_family_index: u32,
) vk.ImageMemoryBarrier2 {
    return .{
        .s_type = .image_memory_barrier_2,
        .p_next = null,
        .src_stage_mask = src.stage_mask,
        .src_access_mask = src.access_mask,
        .dst_stage_mask = dst.stage_mask,
        .dst_access_mask = dst.access_mask,
        .old_layout = src.image_layout,
        .new_layout = dst.image_layout,
        .src_queue_family_index = src_queue_family_index,
        .dst_queue_family_index = dst_queue_family_index,
        .image = handle,
        .subresource_range = .{
            .aspect_mask = subresource.aspect,
            .base_mip_level = subresource.mip_offset,
            .level_count = subresource.mip_count,
            .base_array_layer = subresource.layer_offset,
            .layer_count = subresource.layer_count,
        },
    };
}

/// Same as getImageBarrier() with both queue family indices ignored, which is
/// the default in the C++ signature.
pub fn getImageBarrierSameQueue(
    handle: vk.Image,
    subresource: GPUTextureSubresource,
    src: GPUTextureAccess,
    dst: GPUTextureAccess,
) vk.ImageMemoryBarrier2 {
    return getImageBarrier(
        handle,
        subresource,
        src,
        dst,
        vk.QUEUE_FAMILY_IGNORED,
        vk.QUEUE_FAMILY_IGNORED,
    );
}

pub fn getImageBarrierDependencyInfo(barriers: []const vk.ImageMemoryBarrier2) vk.DependencyInfo {
    return .{
        .s_type = .dependency_info,
        .p_next = null,
        .dependency_flags = .{},
        .memory_barrier_count = 0,
        .p_memory_barriers = null,
        .buffer_memory_barrier_count = 0,
        .p_buffer_memory_barriers = null,
        .image_memory_barrier_count = @intCast(barriers.len),
        .p_image_memory_barriers = barriers.ptr,
    };
}
